engine/notifier_test.go

```diff
@@ -85,123 +84,112 @@ func TestNotifier_ManyNotifications(t *testing.T) {
-		consumedWork := atomic.NewInt32(0)
-
-		// starts the consuming first, because if we starts the production first instead, then
-		// we might finish pushing all jobs, before any of our consumer has started listening
-		// to the queue.
-		var consumersAllReady sync.WaitGroup
-		consumersAllReady.Add(5)
-
-		// 5 routines consuming work
-		for i := 0; i < 5; i++ {
-			go func() {
-				consumersAllReady.Done()
-				for consumedWork.Load() < totalWork {
-					<-notifier.Channel()
-				L:
+		synctest.Test(t, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			notifier := NewNotifier()
+
+			producerCount := int32(10)
+			producerJobs := int32(10)
+			pendingWorkQueue := make(chan struct{}, producerCount*producerJobs)
+			consumedWork := atomic.NewInt32(0)
+
+			// start the consumers first, otherwise we might finish pushing all jobs, before any of
+			// our consumer has started listening to the queue.
+
+			processAllPending := func() {
+				for {
+					select {
+					case <-pendingWorkQueue:
+						consumedWork.Inc()
+					default:
+						return
+					}
+				}
+			}
+
+			// 5 routines consuming work
+			for range 5 {
+				go func() {
 					for {
 						select {
-						case <-pendingWorkQueue:
-							consumedWork.Inc()
-						default:
-							break L
+						case <-ctx.Done():
+							return
+						case <-notifier.Channel():
+							processAllPending()
 						}
 					}
-				}
-			}()
-		}
-
-		// wait long enough for all consumer to be ready for new notification.
-		consumersAllReady.Wait()
-
-		var workersAllReady sync.WaitGroup
-		workersAllReady.Add(10)
-
-		// 10 routines pushing work
-		for i := 0; i < 10; i++ {
-			go func() {
-				workersAllReady.Done()
-				for scheduledWork.Inc() <= totalWork {
-					pendingWorkQueue <- struct{}{}
-					notifier.Notify()
-				}
-			}()
-		}
+				}()
+			}
 
-		// wait long enough for all workers to be started.
-		workersAllReady.Wait()
+			// wait for all consumers to block on the notifier channel
+			synctest.Wait()
 
-		// require that all work is eventually consumed
-		if !conditionEventuallySatisfied(func() bool { return consumedWork.Load() == totalWork }, 3*time.Second, 100*time.Millisecond) {
-			require.Fail(t, "timed out", "only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork)
-		}
-	}
+			for range producerCount {
+				go func() {
+					for range producerJobs {
+						pendingWorkQueue <- struct{}{}
+						notifier.Notify()
+					}
+				}()
+			}
 
-	for r := 0; r < 100; r++ {
-		t.Run(fmt.Sprintf("run %d", r), singleTestRun)
-	}
-}
+			// wait for all producers and consumers to block. at this point, all jobs should be consumed.
+			synctest.Wait()
 
-func conditionEventuallySatisfied(condition func() bool, waitFor time.Duration, tick time.Duration) bool {
-	done := make(chan struct{})
+			// verify all scheduled jobs were consumed.
+			assert.Equal(t, producerCount*producerJobs, consumedWork.Load())
```

Comment:
the comment in line 180 already summarizes what is happening below. 
```suggestion
			// wait for all producers and consumers to block. at this point, all jobs should be consumed.
			synctest.Wait()
			assert.Equal(t, producerCount*producerJobs, consumedWork.Load())
```